File: apps/go/internal/message/message.go
package message

import (
	"regexp"
	"strings"
)

const (
	PlaceholderHint   = "Gebruik {voornaam} of {naam} om het bericht te personaliseren."
	FallbackName      = "student"
	PreviewSampleName = "Jan Jansen"
)

var (
	reVoornaam = regexp.MustCompile(`(?i)\{\s*voornaam\s*\}`)
	reNaam     = regexp.MustCompile(`(?i)\{\s*naam\s*\}`)
)

// FirstName returns the first word of a name, handling "Achternaam, Voornaam" style.
func FirstName(fullName string) string {
	cleaned := strings.TrimSpace(fullName)
	if cleaned == "" {
		return ""
	}
	if idx := strings.Index(cleaned, ","); idx >= 0 {
		after := strings.TrimSpace(cleaned[idx+1:])
		if after != "" {
			cleaned = after
		}
	}
	parts := strings.Fields(cleaned)
	if len(parts) == 0 {
		return ""
	}
	return parts[0]
}

// Personalize replaces {naam}/{voornaam} placeholders (case-insensitive).
func Personalize(message, name string) string {
	if message == "" {
		return message
	}

	full := strings.TrimSpace(name)
	fname := FirstName(full)

	nameRepl := full
	if nameRepl == "" {
		nameRepl = FallbackName
	}
	firstRepl := fname
	if firstRepl == "" {
		if full != "" {
			firstRepl = full
		} else {
			firstRepl = FallbackName
		}
	}

	result := reVoornaam.ReplaceAllString(message, firstRepl)
	result = reNaam.ReplaceAllString(result, nameRepl)
	return result
}

File: apps/python/core/message.py
"""Personalize message templates with contact details."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from core.excel_loader import ContactRow

PLACEHOLDER_HINT = "Gebruik {voornaam} of {naam} om het bericht te personaliseren."

FALLBACK_NAME = "student"
PREVIEW_SAMPLE_NAME = "Jan Jansen"


def first_name(full_name: str) -> str:
    """Return the first word of a name, cleaned up."""
    if not full_name:
        return ""
    cleaned = full_name.strip()
    if not cleaned:
        return ""
    # Handle "Achternaam, Voornaam" style by taking the part after the comma.
    if "," in cleaned:
        cleaned = cleaned.split(",", 1)[1].strip() or cleaned
    return cleaned.split()[0] if cleaned.split() else ""


def personalize(message: str, name: str) -> str:
    """Replace {naam}/{voornaam} placeholders (case-insensitive) in a message."""
    if not message:
        return message

    full = name.strip() if name else ""
    fname = first_name(full)

    def _name_repl(_match: re.Match) -> str:
        return full or FALLBACK_NAME

    def _first_repl(_match: re.Match) -> str:
        return fname or full or FALLBACK_NAME

    result = re.sub(r"\{\s*voornaam\s*\}", _first_repl, message, flags=re.IGNORECASE)
    result = re.sub(r"\{\s*naam\s*\}", _name_repl, result, flags=re.IGNORECASE)
    return result


def personalize_for(message: str, contact: "ContactRow") -> str:
    return personalize(message, getattr(contact, "name", "") or "")
